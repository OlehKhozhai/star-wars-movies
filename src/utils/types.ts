export type Movie = {
  title: string;
  episode_id: number;
  opening_crawl: string;
  director: string;
  producer: string;
  release_date: string;
  created: string;
  url: string;
};

export type FavoriteMovies = Record<string, boolean>;
