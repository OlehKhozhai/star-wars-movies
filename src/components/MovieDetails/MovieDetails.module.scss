.title {
  display: flex;
  justify-content: space-between;
  font-size: 32px;
  margin-bottom: 20px;
}

.starIcon {
  cursor: pointer;
}
