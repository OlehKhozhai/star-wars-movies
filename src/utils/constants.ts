export enum APIPath {
  Films = "https://swapi.dev/api/films",
}


export enum LSKey {
  FavoriteMovies = "FAVORITE_MOVIES",
}