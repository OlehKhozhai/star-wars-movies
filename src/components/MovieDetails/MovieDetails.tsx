import { memo, FC } from "react";

import { StarIcon } from "assets/icons/StarIcon";
import { FavoriteMovies, Movie } from "utils/types";
import styles from "./MovieDetails.module.scss";

type MovieDetailsProps = Movie & {
  favoriteMoviesConfig: FavoriteMovies;
  onFavoriteIconClick: (title: string) => void;
};

const MovieDetails: FC<MovieDetailsProps> = ({
  title,
  opening_crawl,
  favoriteMoviesConfig,
  onFavoriteIconClick,
}) => {
  const handleFavoriteIconClick = () => {
    onFavoriteIconClick(title);
  };

  return (
    <>
      <h1 className={styles.title}>
        {title}
        <span className={styles.starIcon} onClick={handleFavoriteIconClick}>
          <StarIcon fill={favoriteMoviesConfig[title] ? "yellow" : "grey"} />
        </span>
      </h1>

      <p>{opening_crawl}</p>
    </>
  );
};

export default memo(MovieDetails);
