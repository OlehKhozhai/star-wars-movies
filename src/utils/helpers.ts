import { LSKey } from "utils/constants";
import { FavoriteMovies, Movie } from "utils/types";

export const getAPIData = <T>(
  url: string,
  params: RequestInit = {}
): Promise<T> =>
  fetch(url, { ...params })
    .then((data) => data.json())
    .then((data) => data)
    .catch((error) => console.log("--error--", error));

export const initializeFavoriteMoviesConfigToLS = (movies: Movie[]): void => {
  const favoriteMovies = localStorage.getItem(LSKey.FavoriteMovies);

  if (!favoriteMovies) {
    const preparedMovies = movies.reduce<FavoriteMovies>(
      (accumulator, movie) => {
        accumulator[movie.title] = false;

        return accumulator;
      },
      {}
    );

    localStorage.setItem(LSKey.FavoriteMovies, JSON.stringify(preparedMovies));
  }
};

export const getFavoriteMoviesConfigFromLS = (): FavoriteMovies => {
  const favoriteMovies = localStorage.getItem(LSKey.FavoriteMovies);

  return favoriteMovies ? JSON.parse(favoriteMovies) : {};
};

export const setFavoriteMoviesConfigToLS = (config: FavoriteMovies): void => {
  localStorage.setItem(LSKey.FavoriteMovies, JSON.stringify(config));
};

export const updateFavoriteMoviesConfig = (
  title: string
): FavoriteMovies => {
  const favoriteMovies = getFavoriteMoviesConfigFromLS();
  const newFavoriteMovies = {
    ...favoriteMovies,
    [title]: !favoriteMovies[title],
  };

  return newFavoriteMovies;
};
