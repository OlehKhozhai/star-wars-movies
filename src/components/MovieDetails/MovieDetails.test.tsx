import { create } from "react-test-renderer";

import { movieMock } from "utils/mock";
import MovieDetails from "./MovieDetails";

describe("MovieDetails component", () => {
  it("should render by default", () => {
    const component = create(
      <MovieDetails
        {...movieMock}
        favoriteMoviesConfig={{}}
        onFavoriteIconClick={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render favorite movie", () => {
    const component = create(
      <MovieDetails
        {...movieMock}
        favoriteMoviesConfig={{ [movieMock.title]: true }}
        onFavoriteIconClick={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should trigger onFavoriteIconClick function", () => {
    const onFavoriteIconClickStub = jest.fn();
    const component = create(
      <MovieDetails
        {...movieMock}
        favoriteMoviesConfig={{}}
        onFavoriteIconClick={onFavoriteIconClickStub}
      />
    );

    component.root.findByType("span").props.onClick();

    expect(onFavoriteIconClickStub).toHaveBeenCalled();
  });
});
