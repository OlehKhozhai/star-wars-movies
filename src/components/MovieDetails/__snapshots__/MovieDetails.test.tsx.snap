// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails component should render by default 1`] = `
Array [
  <h1
    className="title"
  >
    A New Hope
    <span
      className="starIcon"
      onClick={[Function]}
    >
      <svg
        className=""
        fill="grey"
        height="25"
        width="23"
      >
        <polygon
          points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
          style={
            Object {
              "fillRule": "nonzero",
            }
          }
        />
      </svg>
    </span>
  </h1>,
  <p>
    It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....
  </p>,
]
`;

exports[`MovieDetails component should render favorite movie 1`] = `
Array [
  <h1
    className="title"
  >
    A New Hope
    <span
      className="starIcon"
      onClick={[Function]}
    >
      <svg
        className=""
        fill="yellow"
        height="25"
        width="23"
      >
        <polygon
          points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
          style={
            Object {
              "fillRule": "nonzero",
            }
          }
        />
      </svg>
    </span>
  </h1>,
  <p>
    It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....
  </p>,
]
`;
